### PR TITLE
Setting default `idle_session_lifetime` on tenant import

### PR DESCRIPTION
**Pull request by willvedd**
### 🔧 Changes

Sometimes when importing a `auth0_tenant` resource, the `idle_session_lifetime` field gets set to zero. Normally this isn't such a problem because subsequent plans or applies will have this value set to the default value of 72. However, the Auth0 CLI's `auth0 tf generate` command will get hung up on the zero value because it violates the schema's validation. This default is necessary because not all tenants will have the `idle_session_lifetime` value set, particularly if it is a new tenant.

### 📚 References

- [TF Generate Command - Auth0 CLI](https://auth0.github.io/auth0-cli/auth0_terraform_generate.html)

### 🔬 Testing

Manual verification on a new tenant. Difficult to add to acceptance tests because our testing tenants will always have this value set; it is important to have it _not_ set for testing.

### 📝 Checklist

- [x] All new/changed/fixed functionality is covered by tests (or N/A)
- [x] I have added documentation for all new/changed functionality (or N/A)

### CODE REVIEW

**Review by Widcket — APPROVED** [review]

File: internal/auth0/tenant/flatten.go [hunk 1]
```diff
@@ -26,6 +25,13 @@ func flattenTenant(data *schema.ResourceData, tenant *management.Tenant) error {
 		data.Set("allow_organization_name_in_authentication_api", tenant.GetAllowOrgNameInAuthAPI()),
 	)
 
+	if tenant.GetIdleSessionLifetime() == 0 {
+		idleSessionLifetimeDefault := NewResource().Schema["idle_session_lifetime"].Default
+		result = multierror.Append(result, data.Set("idle_session_lifetime", idleSessionLifetimeDefault))
+	} else {
+		result = multierror.Append(result, data.Set("idle_session_lifetime", tenant.GetIdleSessionLifetime()))
+	}
+
```
Comment (sergiught) [inline internal/auth0/tenant/flatten.go:34]: I would suggest we rework this a little bit as we're opening ourselves up for trouble by directly referencing the schema like this and then the properties' default value. Furthermore the `session_lifetime` is prone to the same issue as it has a default value set.

What about having these defaults as constants references inside the schema and then here. 

Example:
```
// within resource.go

const (
	idleSessionLifetimeDefault = 72
	sessionLifetimeDefault = 168
)

// within this flatten func
func flattenTenant(data *schema.ResourceData, tenant *management.Tenant) error {
	result := multierror.Append(
		data.Set("default_audience", tenant.GetDefaultAudience()),
		data.Set("default_directory", tenant.GetDefaultDirectory()),
		data.Set("default_redirection_uri", tenant.GetDefaultRedirectionURI()),
		data.Set("friendly_name", tenant.GetFriendlyName()),
		data.Set("picture_url", tenant.GetPictureURL()),
		data.Set("support_email", tenant.GetSupportEmail()),
		data.Set("support_url", tenant.GetSupportURL()),
		data.Set("allowed_logout_urls", tenant.GetAllowedLogoutURLs()),
		data.Set("session_lifetime", tenant.GetSessionLifetime()),
		data.Set("idle_session_lifetime", tenant.GetIdleSessionLifetime()),
		data.Set("sandbox_version", tenant.GetSandboxVersion()),
		data.Set("enabled_locales", tenant.GetEnabledLocales()),
		data.Set("flags", flattenTenantFlags(tenant.GetFlags())),
		data.Set("session_cookie", flattenTenantSessionCookie(tenant.GetSessionCookie())),
		data.Set("sessions", flattenTenantSessions(tenant.GetSessions())),
		data.Set("allow_organization_name_in_authentication_api", tenant.GetAllowOrgNameInAuthAPI()),
	)

	if tenant.GetIdleSessionLifetime() == 0 {
		result = multierror.Append(result, data.Set("idle_session_lifetime", idleSessionLifetimeDefault)
	}

	if tenant.GetSessionLifetime() == 0 {
		result = multierror.Append(result, data.Set("session_lifetime", idleSessionLifetimeDefault)
	}

	return result.ErrorOrNil()
}
```

**Review by sergiught — APPROVED** [review]

File: internal/auth0/tenant/flatten.go (hunk 1 above)
Comment (willvedd) [reply internal/auth0/tenant/flatten.go:34]: > we're opening ourselves up for trouble by directly referencing the schema like this and then the properties' default value

What trouble specifically? The default value is defined on the schema, it is _the_ source of truth. We can certainly abstract this value out as a const as a shortcut, but the default value itself should be defined as closely to the schema as possible IMO.

Comment (sergiught) [reply internal/auth0/tenant/flatten.go:34]: We're instantiating the entire resource twice and then referencing a map and it's default value, that can be nil.

File: internal/auth0/tenant/flatten_test.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		tenant := management.Tenant{
+			IdleSessionLifetime: nil,
+			SessionLifetime:     nil,
+		}
+
+		err := flattenTenant(mockResourceData, &tenant)
+
+		assert.NoError(t, err)
+		assert.Equal(t, mockResourceData.Get("idle_session_lifetime"), idleSessionLifetimeDefault)
+		assert.Equal(t, mockResourceData.Get("session_lifetime"), sessionLifetimeDefault)
```
Comment (sergiught) [inline internal/auth0/tenant/flatten_test.go:25]: It's not recommended to reference an implementation constant or var within the test as if these value change erroneously, the tests will still pass. 
```suggestion
		assert.Equal(t, mockResourceData.Get("idle_session_lifetime"), 72)
		assert.Equal(t, mockResourceData.Get("session_lifetime"), 168)
```

File: internal/auth0/tenant/flatten.go [hunk 3]
```diff
@@ -16,8 +16,6 @@ func flattenTenant(data *schema.ResourceData, tenant *management.Tenant) error {
 		data.Set("support_email", tenant.GetSupportEmail()),
 		data.Set("support_url", tenant.GetSupportURL()),
 		data.Set("allowed_logout_urls", tenant.GetAllowedLogoutURLs()),
-		data.Set("session_lifetime", tenant.GetSessionLifetime()),
```
Comment (sergiught) [inline internal/auth0/tenant/flatten.go:19]: We could keep these and then completely remove the elses below.

**Review by sergiught — CHANGES_REQUESTED** [review]

File: internal/auth0/tenant/flatten_test.go (hunk 2 above)
Comment (willvedd) [reply internal/auth0/tenant/flatten_test.go:25]: I'm curious why you think so. These default values are subject to change and the tests assert that they equal the _defaults_, not some arbitrary number.

File: internal/auth0/tenant/flatten.go (hunk 3 above)
Comment (willvedd) [reply internal/auth0/tenant/flatten.go:19]: We could, but we'd potentially be setting the value twice. If an error occurs while setting the first, but the second succeeds, there will still be an error.

File: internal/auth0/tenant/flatten_test.go (hunk 2 above)
Comment (sergiught) [reply internal/auth0/tenant/flatten_test.go:25]: Yes but they could change to a wrong default value and we wouldn't catch that in the tests.

File: internal/auth0/tenant/flatten.go (hunk 3 above)
Comment (sergiught) [reply internal/auth0/tenant/flatten.go:19]: Extremely unlikely that we'd incurr an error considering we're using the safe getters.

Comment (sergiught) [reply internal/auth0/tenant/flatten.go:19]: Also if we don't want to do it twice we can make use of a variable instead. 

```go
sessionLifeTime := tenant.GetSessionLifetime()
if sessionLifeTime == 0 {
  sessionLifeTime = default
}

data.Set(...)
```

**Review by sergiught — APPROVED** [review]

**Comment by willvedd** [comment]
>If idle_session_lifetime and session_lifetime are both 0 for data "auth0_tenant" "this" {} (as well as my imported tenant) does that indicate the value is actually set to 0 or this is part of the problem addressed here?

@dangbert both session lifetime values are required to be positive, non-zero values. The zeros you're seeing are the Go defaults being used in absence of the explicit values being returned from the tenant. In the Auth0 Management API, if the session lifetime values haven't been adjusted, they'll inherit the defaults and not be explicitly defined in your specific tenant object. The tenant that you're exporting to will behave the same as the existing tenant. Though I recommend to review the session lifetime values with your team to make sure they work for your particular business case.

**Comment by dangbert** [comment]
Thanks for the clarification @willvedd. That is indeed consistent with a later discussion I had with my team, so I've omitted the values as you suggested.  Thanks for your fix and guidance, the timing was perfect 👍

**Pull request merged**